lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,639 @@
+		$path_segments = explode( '.', $path );
+		$current       = $store;
+		foreach ( $path_segments as $path_segment ) {
+			if ( isset( $current[ $path_segment ] ) ) {
+				$current = $current[ $path_segment ];
+			} else {
+				return null;
+			}
+		}
+
+		// Returns the opposite if it contains a negator operator (!).
+		return $should_negate_value ? ! $current : $current;
+	}
+
+	/**
+	 * Extracts the directive attribute name to separate and return the directive
+	 * prefix and an optional suffix.
+	 *
+	 * The suffix is the string after the first double hyphen and the prefix is
+	 * everything that comes before the suffix.
+	 *
+	 * Example:
+	 *
+	 *     extract_prefix_and_suffix( 'data-wp-interactive' )   => array( 'data-wp-interactive', null )
+	 *     extract_prefix_and_suffix( 'data-wp-bind--src' )     => array( 'data-wp-bind', 'src' )
+	 *     extract_prefix_and_suffix( 'data-wp-foo--and--bar' ) => array( 'data-wp-foo', 'and--bar' )
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $directive_name The directive attribute name.
+	 * @return array An array containing the directive prefix and optional suffix.
+	 */
+	private function extract_prefix_and_suffix( string $directive_name ): array {
+		return explode( '--', $directive_name, 2 );
+	}
+
+	/**
+	 * Parses and extracts the namespace and reference path from the given
+	 * directive attribute value.
+	 *
+	 * If the value doesn't contain an explicit namespace, it returns the default
+	 * one. If the value contains a JSON instead of a reference path, the function
+	 * tries to parse it and return the resulting array.
+	 *
+	 * Example:
+	 *
+	 *     extract_directive_value( 'actions.foo', 'myPlugin' )                      => array( 'myPlugin', 'actions.foo' )
+	 *     extract_directive_value( 'otherPlugin::actions.foo', 'myPlugin' )         => array( 'otherPlugin', 'actions.foo' )
+	 *     extract_directive_value( '{ "isOpen": false }', 'myPlugin' )              => array( 'myPlugin', array( 'isOpen' => false ) )
+	 *     extract_directive_value( 'otherPlugin::{ "isOpen": false }', 'myPlugin' ) => array( 'otherPlugin', array( 'isOpen' => false ) )
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string|true $directive_value   The directive attribute value. It can be `true` when it's a boolean
+	 *                                       attribute.
+	 * @param string|null $default_namespace Optional. The default namespace if none is explicitly defined.
+	 * @return array An array containing the namespace in the first item and the JSON, the reference path, or null on the
+	 *               second item.
+	 */
+	private function extract_directive_value( $directive_value, $default_namespace = null ): array {
+		if ( empty( $directive_value ) || is_bool( $directive_value ) ) {
+			return array( $default_namespace, null );
+		}
+
+		$matches       = array();
+		$has_namespace = preg_match( '/^([\w\-_\/]+)::(.*)$/', isset( $directive_value ) ? $directive_value : '', $matches );
```

Comment:
to avoid one potential catastrophic case we could end the regex pattern at `::`, since the following `(.*)$` cannot fail to match.

```php
// $directive_value _must_ be set here given the early-abort above, right?
$has_namespace = 1 === preg_match( '~^([\w/_-]+)::~', $directive_value, $namespace_match );
if ( $has_namespace ) {
	$default_namespace = $matches[1];
	$directive_value   = substr( $directive_value, strlen( $matches[0] ) );
}

$might_be_json_object = '{' === $directive_value[0];
if ( $might_be_json_object ) {
	$decoded = json_decode( $directive_value );

	// Should we use an affirmative test here like `is_array( $decoded )`
	if ( isset( $decoded ) && 'null' !== $decoded ) {
		$directive_value = $decoded;
	}
}

return array( $default_namespace, $directive_value );
```

Reply:
Improved in https://github.com/WordPress/gutenberg/pull/58066/commits/560a17ec69a7845fd8d66f4b1cf6b51a9786e831.

I didn't add this because we also rely on `json_decode` to turn boolean and number strings into booleans and numbers.

`$might_be_json_object = '{' === $directive_value[0];`